File: fullstacks/rails/app/services/llm_client.rb
require "net/http"
require "json"
require "uri"

# 结构化 JSON 生成的 LLM 客户端。日志规范、网关重试、CF 1010 等坑见 docs/dev-notes.md §3。
# 契约测试默认 LLM_ENABLED=false 时立即抛异常，由上层走本地兜底；HTTP 路径不在测试链上。
module LlmClient
  module_function

  class Error < StandardError
    attr_reader :status

    def initialize(message, status: 0)
      super(message)
      @status = status
    end
  end

  TRANSIENT = [
    Timeout::Error, Errno::ECONNRESET, Errno::ECONNREFUSED, Errno::EPIPE,
    EOFError, SocketError, IOError, OpenSSL::SSL::SSLError
  ].freeze

  def generate_capsule_suggestion(prompt)
    generate_structured_json(prompt, suggestion_spec)
  end

  def generate_capsule_recommendations(prompt)
    generate_structured_json(prompt, recommendation_spec)
  end

  def generate_structured_json(prompt, spec)
    raise Error.new("LLM is disabled or missing API key") if !AppConfig.llm_enabled || AppConfig.llm_api_key.empty?

    case AppConfig.llm_api_style
    when "responses" then generate_with_responses(prompt, spec)
    when "auto"
      begin
        generate_with_responses(prompt, spec)
      rescue Error => e
        Rails.logger.info("Responses API unavailable (#{e.message}); falling back to chat completions")
        generate_with_chat(prompt, spec, disable_thinking: true)
      end
    else
      generate_with_chat(prompt, spec, disable_thinking: true)
    end
  end

  def generate_with_responses(prompt, spec)
    payload = {
      model: AppConfig.llm_model,
      input: prompt,
      max_output_tokens: spec[:max_output_tokens],
      text: { format: { type: "json_schema", name: spec[:schema_name], strict: true, schema: spec[:schema] } },
    }
    parse_json_object(extract_responses_text(post_json(responses_url, payload)))
  end

  def generate_with_chat(prompt, spec, disable_thinking:)
    build = lambda do |thinking|
      p = {
        model: AppConfig.llm_model,
        messages: [
          { role: "system", content: spec[:system_prompt] },
          { role: "user", content: prompt },
        ],
        max_tokens: spec[:max_tokens],
      }
      p[:thinking] = { type: "disabled" } if thinking
      p
    end
    begin
      parse_json_object(extract_chat_text(post_json(chat_url, build.call(disable_thinking))))
    rescue Error => e
      raise e unless e.status == 400

      # 某些网关不认 thinking 字段，去掉重试一次。
      parse_json_object(extract_chat_text(post_json(chat_url, build.call(false))))
    end
  end

  # 向 url POST JSON；瞬时网络/TLS 错误按配置重试，HTTP 4xx/5xx 与坏 JSON 不重试。
  def post_json(url, payload)
    body = JSON.generate(payload)
    attempts = [1, AppConfig.llm_max_retries + 1].max
    last_error = nil
    uri = URI.parse(url)

    (1..attempts).each do |attempt|
      Rails.logger.info("LLM request  model=#{AppConfig.llm_model} url=#{url} attempt=#{attempt}/#{attempts}")
      start = now_ms
      begin
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = uri.scheme == "https"
        http.open_timeout = AppConfig.llm_timeout_ms / 1000.0
        http.read_timeout = AppConfig.llm_timeout_ms / 1000.0
        req = Net::HTTP::Post.new(uri)
        req["Authorization"] = "Bearer #{AppConfig.llm_api_key}"
        req["Content-Type"] = "application/json"
        req["Accept"] = "application/json"
        req["User-Agent"] = AppConfig.llm_user_agent
        req.body = body
        res = http.request(req)
        elapsed = now_ms - start
        status = res.code.to_i
        if status < 200 || status >= 300
          Rails.logger.warn("LLM error    model=#{AppConfig.llm_model} elapsed_ms=#{elapsed} status=#{status}")
          raise Error.new("HTTP #{status}: #{res.body}", status: status)
        end
        parsed = begin
          JSON.parse(res.body)
        rescue JSON::ParserError => e
          Rails.logger.warn("LLM error    model=#{AppConfig.llm_model} elapsed_ms=#{elapsed} error=invalid-json")
          raise Error.new("LLM response was not valid JSON")
        end
        Rails.logger.info("LLM response model=#{AppConfig.llm_model} elapsed_ms=#{elapsed} tokens=#{extract_tokens(parsed)}")
        return parsed
      rescue Error
        raise
      rescue *TRANSIENT => e
        elapsed = now_ms - start
        will_retry = attempt < attempts
        Rails.logger.warn("LLM error    model=#{AppConfig.llm_model} elapsed_ms=#{elapsed} error=#{e.message}#{will_retry ? ' (will retry)' : ''}")
        last_error = e
        sleep(AppConfig.llm_retry_backoff_ms * attempt / 1000.0) if will_retry
      end
    end
    raise Error.new(last_error&.message || "LLM request failed")
  end

  def extract_tokens(body)
    usage = body["usage"]
    return "n/a" unless usage.is_a?(Hash)

    total = usage["total_tokens"].to_i
    return total.to_s if total.positive?

    sum = usage["input_tokens"].to_i + usage["output_tokens"].to_i
    sum.positive? ? sum.to_s : "n/a"
  end

  def extract_chat_text(body)
    choices = body["choices"]
    raise Error.new("LLM chat response missing choices") unless choices.is_a?(Array) && choices.any?

    content = choices.dig(0, "message", "content")
    raise Error.new("LLM chat response missing content") if content.nil? || content.to_s.strip.empty?

    content
  end

  def extract_responses_text(body)
    ot = body["output_text"]
    return ot if ot.is_a?(String) && !ot.strip.empty?

    Array(body["output"]).each do |item|
      Array(item.is_a?(Hash) ? item["content"] : nil).each do |entry|
        txt = entry.is_a?(Hash) ? entry["text"] : nil
        return txt if txt.is_a?(String) && !txt.strip.empty?
      end
    end
    raise Error.new("LLM response did not contain output text")
  end

  def parse_json_object(raw)
    text = raw.to_s.strip
    if text.start_with?("```")
      text = text.sub(/\A```[a-zA-Z]*\s*/, "").sub(/\s*```\z/, "").strip
    end
    begin
      JSON.parse(text)
    rescue JSON::ParserError
      s = text.index("{")
      e = text.rindex("}")
      raise Error.new("LLM output was not valid JSON") if s.nil? || e.nil? || e <= s

      JSON.parse(text[s..e])
    end
  end

  def responses_url = "#{strip_slash(AppConfig.llm_base_url)}/responses"
  def chat_url = "#{strip_slash(AppConfig.llm_base_url)}/chat/completions"

  def strip_slash(v)
    v.to_s.strip.empty? ? "https://api.openai.com/v1" : v.sub(%r{/+\z}, "")
  end

  def now_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) * 1000).to_i

  def suggestion_spec
    {
      schema_name: "capsule_suggestion",
      schema: {
        type: "object", additionalProperties: false,
        required: %w[title content openInDays],
        properties: {
          title: { type: "string" }, content: { type: "string" },
          openInDays: { type: "integer", minimum: 1, maximum: 3650 },
        },
      },
      system_prompt: "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。" \
                     "JSON 必须包含字符串字段 title、content 和整数字段 openInDays。若用户已给出标题，title 可原样回填。",
      max_output_tokens: 900, max_tokens: 900,
    }
  end

  def recommendation_spec
    {
      schema_name: "capsule_recommendations",
      schema: {
        type: "object", additionalProperties: false, required: %w[items],
        properties: {
          items: {
            type: "array", minItems: 3, maxItems: 8,
            items: {
              type: "object", additionalProperties: false,
              required: %w[title hint openInDays],
              properties: {
                title: { type: "string" }, hint: { type: "string" },
                openInDays: { type: "integer", minimum: 1, maximum: 3650 },
              },
            },
          },
        },
      },
      system_prompt: "你只返回严格 JSON 对象，不要 Markdown、代码块或解释。" \
                     "JSON 必须包含数组字段 items，每项含字符串字段 title、hint 和整数字段 openInDays。",
      max_output_tokens: 900, max_tokens: 900,
    }
  end
end

File: fullstacks/laravel/app/Services/LlmClient.php
<?php

namespace App\Services;

use App\Exceptions\LlmException;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;
use Throwable;

/**
 * 结构化 JSON 生成的 LLM 客户端。日志规范、网关重试、chat 风格跳过 /responses 等坑见 docs/dev-notes.md。
 * 契约测试默认 LLM_ENABLED=false 时立即抛异常，由上层走本地兜底；HTTP 路径不在测试链上。
 *
 * 日志规范（CLAUDE.md）：请求前 INFO `LLM request`、成功 INFO `LLM response`、失败 WARNING `LLM error`。
 */
class LlmClient
{
    /**
     * 读配置：优先 getenv()（直接读进程环境，可靠），回退 Laravel env()。
     * `php artisan serve` 的 worker 在 variables_order 不含 E 时，仅继承（非 .env）的变量
     * 可能不进 $_SERVER/$_ENV，导致 env() 读不到；getenv() 不受此限。
     */
    private function conf(string $key, string $default): string
    {
        $value = getenv($key);
        if ($value === false || $value === '') {
            $value = env($key);
        }
        return ($value === false || $value === null || $value === '') ? $default : (string)$value;
    }

    public function enabled(): bool
    {
        return strtolower($this->conf('LLM_ENABLED', 'false')) === 'true' && $this->apiKey() !== '';
    }

    public function provider(): string
    {
        return $this->conf('LLM_PROVIDER', 'openai');
    }

    public function model(): string
    {
        return $this->conf('LLM_MODEL', 'gpt-4.1-mini');
    }

    private function apiKey(): string
    {
        return $this->conf('LLM_API_KEY', '');
    }

    private function baseUrl(): string
    {
        return rtrim($this->conf('LLM_BASE_URL', 'https://api.openai.com/v1'), '/');
    }

    private function apiStyle(): string
    {
        return $this->conf('LLM_API_STYLE', 'chat');
    }

    private function timeoutMs(): int
    {
        return (int)$this->conf('LLM_TIMEOUT_MS', '30000');
    }

    private function maxRetries(): int
    {
        return (int)$this->conf('LLM_MAX_RETRIES', '2');
    }

    private function backoffMs(): int
    {
        return (int)$this->conf('LLM_RETRY_BACKOFF_MS', '400');
    }

    private function userAgent(): string
    {
        return $this->conf(
            'LLM_USER_AGENT',
            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
        );
    }

    public function generateCapsuleSuggestion(string $prompt): array
    {
        return $this->generateStructuredJson($prompt, $this->suggestionSpec());
    }

    public function generateCapsuleRecommendations(string $prompt): array
    {
        return $this->generateStructuredJson($prompt, $this->recommendationSpec());
    }

    private function generateStructuredJson(string $prompt, array $spec): array
    {
        if (!$this->enabled()) {
            throw new LlmException('LLM is disabled or missing API key');
        }

        switch ($this->apiStyle()) {
            case 'responses':
                return $this->generateWithResponses($prompt, $spec);
            case 'auto':
                try {
                    return $this->generateWithResponses($prompt, $spec);
                } catch (LlmException $e) {
                    Log::info("Responses API unavailable ({$e->getMessage()}); falling back to chat completions");
                    return $this->generateWithChat($prompt, $spec, true);
                }
            default:
                return $this->generateWithChat($prompt, $spec, true);
        }
    }

    private function generateWithResponses(string $prompt, array $spec): array
    {
        $payload = [
            'model' => $this->model(),
            'input' => $prompt,
            'max_output_tokens' => $spec['max_tokens'],
            'text' => ['format' => ['type' => 'json_schema', 'name' => $spec['schema_name'], 'strict' => true, 'schema' => $spec['schema']]],
        ];
        return $this->parseJsonObject($this->extractResponsesText($this->postJson($this->baseUrl() . '/responses', $payload)));
    }

    private function generateWithChat(string $prompt, array $spec, bool $disableThinking): array
    {
        $build = function (bool $thinking) use ($prompt, $spec): array {
            $payload = [
                'model' => $this->model(),
                'messages' => [
                    ['role' => 'system', 'content' => $spec['system_prompt']],
                    ['role' => 'user', 'content' => $prompt],
                ],
                'max_tokens' => $spec['max_tokens'],
            ];
            if ($thinking) $payload['thinking'] = ['type' => 'disabled'];
            return $payload;
        };

        try {
            return $this->parseJsonObject($this->extractChatText($this->postJson($this->baseUrl() . '/chat/completions', $build($disableThinking))));
        } catch (LlmException $e) {
            if ($e->status !== 400) throw $e;
            // 某些网关不认 thinking 字段，去掉重试一次。
            return $this->parseJsonObject($this->extractChatText($this->postJson($this->baseUrl() . '/chat/completions', $build(false))));
        }
    }

    /** 向 url POST JSON；瞬时网络/TLS 错误按配置重试，HTTP 4xx/5xx 与坏 JSON 不重试。 */
    private function postJson(string $url, array $payload): array
    {
        $attempts = max(1, $this->maxRetries() + 1);
        $lastError = null;

        for ($attempt = 1; $attempt <= $attempts; $attempt++) {
            Log::info("LLM request  model={$this->model()} url={$url} attempt={$attempt}/{$attempts}");
            $start = hrtime(true);
            try {
                $response = Http::withHeaders([
                    'Authorization' => 'Bearer ' . $this->apiKey(),
                    'Content-Type' => 'application/json',
                    'Accept' => 'application/json',
                    'User-Agent' => $this->userAgent(),
                ])
                    ->connectTimeout((int)ceil($this->timeoutMs() / 1000))
                    ->timeout((int)ceil($this->timeoutMs() / 1000))
                    ->withBody(json_encode($payload, JSON_UNESCAPED_UNICODE), 'application/json')
                    ->post($url);

                $elapsed = $this->elapsedMs($start);
                $status = $response->status();
                if ($status < 200 || $status >= 300) {
                    Log::warning("LLM error    model={$this->model()} elapsed_ms={$elapsed} status={$status}");
                    throw new LlmException("HTTP {$status}: " . $response->body(), $status);
                }

                $parsed = json_decode($response->body(), true);
                if (!is_array($parsed)) {
                    Log::warning("LLM error    model={$this->model()} elapsed_ms={$elapsed} error=invalid-json");
                    throw new LlmException('LLM response was not valid JSON');
                }

                Log::info("LLM response model={$this->model()} elapsed_ms={$elapsed} tokens={$this->extractTokens($parsed)}");
                return $parsed;
            } catch (LlmException $e) {
                // HTTP / JSON 错误：不重试。
                throw $e;
            } catch (Throwable $e) {
                $elapsed = $this->elapsedMs($start);
                $willRetry = $attempt < $attempts;
                Log::warning("LLM error    model={$this->model()} elapsed_ms={$elapsed} error={$e->getMessage()}" . ($willRetry ? ' (will retry)' : ''));
                $lastError = $e;
                if ($willRetry) usleep($this->backoffMs() * $attempt * 1000);
            }
        }

        throw new LlmException($lastError?->getMessage() ?? 'LLM request failed');
    }

    private function extractTokens(array $body): string
    {
        $usage = $body['usage'] ?? null;
        if (!is_array($usage)) return 'n/a';
        $total = (int)($usage['total_tokens'] ?? 0);
        if ($total > 0) return (string)$total;
        $sum = (int)($usage['input_tokens'] ?? 0) + (int)($usage['output_tokens'] ?? 0);
        return $sum > 0 ? (string)$sum : 'n/a';
    }

    private function extractChatText(array $body): string
    {
        $choices = $body['choices'] ?? null;
        if (!is_array($choices) || $choices === []) throw new LlmException('LLM chat response missing choices');
        $content = $choices[0]['message']['content'] ?? null;
        if (!is_string($content) || trim($content) === '') throw new LlmException('LLM chat response missing content');
        return $content;
    }

    private function extractResponsesText(array $body): string
    {
        $ot = $body['output_text'] ?? null;
        if (is_string($ot) && trim($ot) !== '') return $ot;
        foreach (($body['output'] ?? []) as $item) {
            foreach ((is_array($item) ? ($item['content'] ?? []) : []) as $entry) {
                $txt = is_array($entry) ? ($entry['text'] ?? null) : null;
                if (is_string($txt) && trim($txt) !== '') return $txt;
            }
        }
        throw new LlmException('LLM response did not contain output text');
    }

    private function parseJsonObject(string $raw): array
    {
        $text = trim($raw);
        if (str_starts_with($text, '```')) {
            $text = trim(preg_replace('/\A```[a-zA-Z]*\s*/', '', preg_replace('/\s*```\z/', '', $text)));
        }
        $decoded = json_decode($text, true);
        if (is_array($decoded)) return $decoded;

        $start = strpos($text, '{');
        $end = strrpos($text, '}');
        if ($start === false || $end === false || $end <= $start) {
            throw new LlmException('LLM output was not valid JSON');
        }
        $decoded = json_decode(substr($text, $start, $end - $start + 1), true);
        if (!is_array($decoded)) throw new LlmException('LLM output was not valid JSON');
        return $decoded;
    }

    private function elapsedMs(int $startNs): int
    {
        return (int)((hrtime(true) - $startNs) / 1_000_000);
    }

    private function suggestionSpec(): array
    {
        return [
            'schema_name' => 'capsule_suggestion',
            'schema' => [
                'type' => 'object',
                'additionalProperties' => false,
                'required' => ['title', 'content', 'openInDays'],
                'properties' => [
                    'title' => ['type' => 'string'],
                    'content' => ['type' => 'string'],
                    'openInDays' => ['type' => 'integer', 'minimum' => 1, 'maximum' => 3650],
                ],
            ],
            'system_prompt' => '你只返回严格 JSON 对象，不要 Markdown、代码块或解释。'
                . 'JSON 必须包含字符串字段 title、content 和整数字段 openInDays。若用户已给出标题，title 可原样回填。',
            'max_tokens' => 900,
        ];
    }

    private function recommendationSpec(): array
    {
        return [
            'schema_name' => 'capsule_recommendations',
            'schema' => [
                'type' => 'object',
                'additionalProperties' => false,
                'required' => ['items'],
                'properties' => [
                    'items' => [
                        'type' => 'array',
                        'minItems' => 3,
                        'maxItems' => 8,
                        'items' => [
                            'type' => 'object',
                            'additionalProperties' => false,
                            'required' => ['title', 'hint', 'openInDays'],
                            'properties' => [
                                'title' => ['type' => 'string'],
                                'hint' => ['type' => 'string'],
                                'openInDays' => ['type' => 'integer', 'minimum' => 1, 'maximum' => 3650],
                            ],
                        ],
                    ],
                ],
            ],
            'system_prompt' => '你只返回严格 JSON 对象，不要 Markdown、代码块或解释。'
                . 'JSON 必须包含数组字段 items，每项含字符串字段 title、hint 和整数字段 openInDays。',
            'max_tokens' => 900,
        ];
    }
}
